cmd/healthcheck-handler.go

```diff
@@ -17,31 +17,36 @@
 package cmd
 
 import (
-	"fmt"
 	"net/http"
 	"os"
-	"runtime"
 
 	xhttp "github.com/minio/minio/cmd/http"
 	"github.com/minio/minio/cmd/logger"
 )
 
-const (
-	minioHealthGoroutineThreshold = 10000
-)
-
-// ReadinessCheckHandler -- checks if there are more than threshold
-// number of goroutines running, returns service unavailable.
-//
-// Readiness probes are used to detect situations where application
-// is under heavy load and temporarily unable to serve. In a orchestrated
-// setup like Kubernetes, containers reporting that they are not ready do
-// not receive traffic through Kubernetes Services.
+// ReadinessCheckHandler -- Checks if the quorum number of disks are available.
+// For FS - Checks if the backend disk is available
+// For Zones - Checks if all the zones have enough quorum
 func ReadinessCheckHandler(w http.ResponseWriter, r *http.Request) {
-	if err := goroutineCountCheck(minioHealthGoroutineThreshold); err != nil {
+	ctx := newContext(r, w, "ReadinessCheckHandler")
+
+	objLayer := newObjectLayerFn()
+	// Service not initialized yet
+	if objLayer == nil {
+		// Respond with 200 OK while server initializes to ensure a distributed cluster
+		// is able to start on orchestration platforms like Docker Swarm.
```

Comment:
Kubernetes has had `publishNotReadyAddresses` in Services since 1.8, which allows the creation of an internal service that makes it possible for all nodes to discover and communicate with each other, even if they are not ready. Posting ready here will cause a service that expects all nodes to actually be ready to return random "Server not initialized, please try again" responses.

NotReady (starting) -> Ready (initializing) -> NotReady (no quorum) -> Ready (quorum) seems to mix two definitions of ready: ready to accept client requests and ready to respond in general. I think these should be split into readiness to accept connections and readiness to serve clients.

Should this be a separate issue?

Reply:
IMO we should send `http.StatusServiceUnavailable` here as well. Since no disks are ready to accept any requests. Will explore more on this

Reply:
That is intentionally kept as 200 ok for routing requests @Praveenrajmani check the original PR for why it's done this way.

Reply:
That is not necessary for readiness IMO @Praveenrajmani so you should return 503 here.

Reply:
Agreed @harshavardhana